service/strong_consistency/sc_state_machine.cc

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (C) 2025-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: LicenseRef-ScyllaDB-Source-Available-1.0
+ */
+
+#include "sc_state_machine.hh"
+#include "serializer_impl.hh"
+#include "idl/sc_state_machine.dist.hh"
+#include "idl/sc_state_machine.dist.impl.hh"
+#include "replica/database.hh"
+
+namespace service {
+    static logging::logger logger("sc_sm");
+
+    class sc_state_machine: public raft_state_machine {
+        locator::global_tablet_id _tablet;
+        raft::group_id _group_id;
+        replica::database& _db;
+
+    public:
+        sc_state_machine(locator::global_tablet_id tablet,
+            raft::group_id gid,
+            replica::database& db)
+            : _tablet(tablet)
+            , _group_id(gid)
+            , _db(db)
+        {
+        }
+
+        future<> apply(std::vector<raft::command_cref> command) override {
+            try {
+                utils::chunked_vector<frozen_mutation> muts;
+                muts.reserve(command.size());
+                for (const auto& c: command) {
+                    auto is = ser::as_input_stream(c);
+                    const auto cmd = ser::deserialize(is, std::type_identity<sc_raft_command>{});
+                    muts.push_back(std::move(cmd.mutation));
+                }
+
+                // FIXME: fix db.apply atomicity
```

Comment:
But why do we need atomicity between different commands given that each command represents exactly one write?

Reply:
We can't publish the second command to the memtable while the first one is not published yet, can we?

Reply:
We want to batch-write many commands to the commitlog, `database::do_apply_many` achieves that.

Probably we can implement our own custom "sequential apply" in `sc_state_machine`:
* write all commands to the commitlog in one batch, like `do_apply_many` does
* call `apply_in_memory` sequentially

=> no need to guarantee memory for the entire batch, so that `run_when_memory_available` doesn't yield